Add WKWebsiteDataStore SPI to fetch Local Storage data

RupinMittal opened the pull request:
#### 5268730b8463f96170dfe8dbbcb3badf1d3bc3ec
<pre>
Add WKWebsiteDataStore SPI to fetch Local Storage data
<a href="https://bugs.webkit.org/show_bug.cgi?id=283593">https://bugs.webkit.org/show_bug.cgi?id=283593</a>
<a href="https://rdar.apple.com/140440709">rdar://140440709</a>

Reviewed by Sihui Liu.

Some clients may want to restore the local storage of ephemeral
data stores. (For example, Safari restoring the local storage of
private browsing tabs).

The restoration process will have two parts:
1. Client fetches the local storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

This patch adds an SPI (which will eventually go through API
review) that does the first part of the restoration (fetch).

The SPI works for both ephemeral and persistent data stores,
so there is an API test for each.

The SPI also works for both first party storage and for third party
storage, so there are API tests for each.

For now the tests simply check that the returned NSData is not null.
But once we add restoration as well, we&apos;ll modify the tests to test
that the data returned is the correct data too.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchLocalStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::fetchStorageMap const):
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm: Added.
(testFetchLocalStorage):
(TEST(WebKit, FetchLocalStorageFromPersistentDataStore)):
(TEST(WebKit, FetchLocalStorageFromEphemeralDataStore)):
(-[RestoreLocalStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(postMessage):
(setItem):
(FetchLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(FetchLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):

Canonical link: <a href="https://commits.webkit.org/287349@main">https://commits.webkit.org/287349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87e49ab537caef092ff381ec73934b8b7d3098e7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79171 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83795 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30359 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81304 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6468 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61941 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19853 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82238 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42246 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49341 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26273 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28733 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85189 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6472 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4482 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html media/video-replaces-poster.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70185 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6635 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68016 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69434 "88 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17327 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13479 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12293 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12228 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6370 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8162 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
```diff
@@ -65,6 +65,8 @@ class LocalStorageManager {
     void cancelConnectToTransientLocalStorageArea(IPC::Connection::UniqueID);
     void disconnectFromStorageArea(IPC::Connection::UniqueID, StorageAreaIdentifier);
 
+    HashMap<String, String> fetchStorageMap();
```
`HashMap<String, String> fetchStorageMap() const;` ?

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -121,6 +121,7 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver, public
     void resume();
     void handleLowMemoryWarning();
     void syncLocalStorage(CompletionHandler<void()>&&);
+    void fetchLocalStorage(CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&&);
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1426,6 +1427,23 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler
+{
+    if ([dataTypes containsObject:WKLocalStorage]) {
+        _websiteDataStore->fetchLocalStorage([completionHandler = makeBlockPtr(completionHandler)](HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorage) {
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
```diff
@@ -155,6 +155,11 @@ typedef NS_ENUM(uint8_t, _WKRestrictedOpenerType) {
 - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
 
+/*! @constant WKLocalStorage Local Storage. */
+WK_EXTERN NSString * const WKLocalStorage WK_API_AVAILABLE(macos(WK_MAC_TBA));
```
Can't we just rely on the existing `WKWebsiteDataTypeLocalStorage` ?

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -2001,6 +2001,16 @@ void NetworkProcessProxy::setEmulatedConditions(PAL::SessionID sessionID, std::o
 
 #endif // ENABLE(INSPECTOR_NETWORK_THROTTLING)
 
+void NetworkProcessProxy::fetchLocalStorage(PAL::SessionID sessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&& completionHandler)
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1426,6 +1427,23 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+    if ([dataTypes containsObject:WKLocalStorage]) {
+        _websiteDataStore->fetchLocalStorage([completionHandler = makeBlockPtr(completionHandler)](HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorage) {
+            WTF::Persistence::Encoder encoder;
+            encoder << localStorage;
```
We should encode a version number first. Otherwise, we'll have trouble with backwards compatibility in the future.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1426,6 +1427,23 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
+        _websiteDataStore->fetchLocalStorage([completionHandler = makeBlockPtr(completionHandler)](HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorage) {
+            WTF::Persistence::Encoder encoder;
+            encoder << localStorage;
+            completionHandler([[NSData alloc] initWithBytes:encoder.buffer() length:encoder.bufferSize()]);
```
The better way to do this in to include `<wtf/cocoa/SpanCocoa.h>` and use `toNSData(encoder.span())`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -0,0 +1,97 @@
+
+TEST(WebKit, RestoreLocalStorageEphemeralDataStore)
+{
+    RetainPtr storeConfiguration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] initNonPersistentConfiguration]);
```
Please move the implementation to a static function and have both tests call the same function (with a different data store parameter), to avoid code duplication.

RupinMittal replied:
I made the parameter webView since we need it for some of function calls. Is this ok?

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -3102,4 +3102,15 @@ void NetworkProcess::setPersistedDomains(PAL::SessionID sessionID, HashSet<Regis
         session->setPersistedDomains(WTFMove(domains));
 }
 
+void NetworkProcess::fetchLocalStorage(PAL::SessionID sessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&& completionHandler)
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&`

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -450,6 +450,8 @@ class NetworkProcess final : public AuxiliaryProcess, private DownloadManager::C
 
     bool enableModernDownloadProgress() const { return m_enableModernDownloadProgress; }
 
+    void fetchLocalStorage(PAL::SessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>)>&&);
```
`HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&`

cdumez commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
```diff
@@ -65,6 +65,8 @@ class LocalStorageManager {
     void cancelConnectToTransientLocalStorageArea(IPC::Connection::UniqueID);
     void disconnectFromStorageArea(IPC::Connection::UniqueID, StorageAreaIdentifier);
 
+    HashMap<String, String> const fetchStorageMap();
```
`HashMap<String, String> fetchStorageMap() const`?

The `const` is at a weird location..

RupinMittal replied:
Oops! Fixed.

RupinMittal commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1259,6 +1259,27 @@ void NetworkStorageManager::syncLocalStorage(CompletionHandler<void()>&& complet
+        }
+
+        RunLoop::protectedMain()->dispatch([completionHandler = WTFMove(completionHandler), localStorageMap = crossThreadCopy(WTFMove(localStorageMap))] mutable {
+            completionHandler(WTFMove(localStorageMap));
```
Is the `crossThreadCopy()` needed here? I added it because we're moving the HashMap to a different thread--but if we're using `WTFMove()` and not using the HashMap on this work thread afterwards, do we still need the `crossThreadCopy()`?

szewai replied:
Yes; `crossThreadCopy` knows to avoid copy if possible, and it should be always invoked.

szewai commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -3102,4 +3102,15 @@ void NetworkProcess::setPersistedDomains(PAL::SessionID sessionID, HashSet<Regis
 
+void NetworkProcess::fetchLocalStorage(PAL::SessionID sessionID, CompletionHandler<void(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&&)>&& completionHandler)
+{
+    auto* session = networkSession(sessionID);
```
`CheckedPtr`

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1259,6 +1259,27 @@ void NetworkStorageManager::syncLocalStorage(CompletionHandler<void()>&& complet
+
+        HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorageMap;
+
+        for (auto& [clientOrigin, manager] : m_originStorageManagers) {
```
`m_originStorageManager` only tracks the origins that are active. Technically we should use `getAllOrigins()` for all data.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1259,6 +1259,27 @@ void NetworkStorageManager::syncLocalStorage(CompletionHandler<void()>&& complet
+        HashMap<WebCore::ClientOrigin, HashMap<String, String>> localStorageMap;
+
+        for (auto& [clientOrigin, manager] : m_originStorageManagers) {
+            if (auto* localStorageManager = manager->existingLocalStorageManager())
```
Ditto, this should be `originStorageManager(origin, ShouldWriteOriginFile::No).localStorageManager(*m_storageAreaRegistry)`

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1259,6 +1259,27 @@ void NetworkStorageManager::syncLocalStorage(CompletionHandler<void()>&& complet
+
+        for (auto& [clientOrigin, manager] : m_originStorageManagers) {
+            if (auto* localStorageManager = manager->existingLocalStorageManager())
+                localStorageMap.add(clientOrigin, localStorageManager->fetchStorageMap());
```
We should not add `ClientOrigin` to result if map is empty.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -0,0 +1,79 @@
+    Util::run(&doneFetching);
+}
+
+TEST(WebKit, RestoreLocalStoragePersistentDataStore)
```
The test name is confusing: there is no restoration here.

RupinMittal replied:
I named the tests that way because the next patch will add the restoration part and these tests will be expanded to test that. Is that ok?

szewai replied:
You could add another unit test to this file for restoring data when the restore SPI is available; the current test is about fetching data only.

szewai commented on Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
```diff
@@ -243,4 +243,9 @@ void LocalStorageManager::disconnectFromStorageArea(IPC::Connection::UniqueID co
 
+HashMap<String, String> LocalStorageManager::fetchStorageMap() const
+{
+    return m_localStorageArea->allItems();
```
The LocalStorage items will be in either `m_localStorageArea` (first-party) **or** 
`m_transientStorageArea` (third-party), so you should null check `m_localStorageArea` and try getting from `m_transientStorageArea ` if `m_localStorageArea` is null.
And `m_transientStorageArea` is also pointer, so you need null check for `m_transientStorageArea` too.

To ensure the implementation is correct, please add a test for localStorage in third-party iframe. For how to add 3rd frame for API test, you might check FileSystemAccess.FetchDataForThirdParty.

RupinMittal replied:
Added!

szewai commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -2001,6 +2001,16 @@ void NetworkProcessProxy::setEmulatedConditions(PAL::SessionID sessionID, std::o
+    if (!canSendMessage()) {
+        completionHandler({ });
+        return;
+    }
```
Not needed

RupinMittal replied:
If we don't return from here, won't we unnecessarily send the IPC to the Network Process?

szewai replied:
If message cannot be sent, `sendWithAsyncReply` will not send message (`AuxiliaryProcessProxy::sendMessage`).

RupinMittal replied:
Oh right, that makes sense.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2710,4 +2710,15 @@ void WebsiteDataStore::setRestrictedOpenerTypeForDomainForTesting(const WebCore:
+        return;
+    }
+
+    networkProcessProxy->fetchLocalStorage(m_sessionID, WTFMove(completionHandler));
```
We will need to launch network process to fetch the data:
`protectedNetworkProcess()->fetchLocalStorage(m_sessionID, WTFMove(completionHandler));`

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm:
```diff
@@ -0,0 +1,178 @@
+    Ephemeral
+};
+
+static void testRestoreLocalStorageThirdPartyIFrame(DataStoreType dataStoreType)
```
This function can directly take `RetainPtr<WKWebsiteDataStore>`, and you could construct store before it; no need for the enum

RupinMittal committed:
Add WKWebsiteDataStore SPI to fetch Local Storage data
https://bugs.webkit.org/show_bug.cgi?id=283593
rdar://140440709

Reviewed by Sihui Liu.

Some clients may want to restore the local storage of ephemeral
data stores. (For example, Safari restoring the local storage of
private browsing tabs).

The restoration process will have two parts:
1. Client fetches the local storage data from WebKit and holds onto it.
2. Client gives back the data to WebKit to restore it.

This patch adds an SPI (which will eventually go through API
review) that does the first part of the restoration (fetch).

The SPI works for both ephemeral and persistent data stores,
so there is an API test for each.

The SPI also works for both first party storage and for third party
storage, so there are API tests for each.

For now the tests simply check that the returned NSData is not null.
But once we add restoration as well, we'll modify the tests to test
that the data returned is the correct data too.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchLocalStorage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.cpp:
(WebKit::LocalStorageManager::fetchStorageMap const):
* Source/WebKit/NetworkProcess/storage/LocalStorageManager.h:
* Source/WebKit/NetworkProcess/storage/MemoryStorageArea.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::fetchLocalStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _fetchDataOfTypes:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::fetchLocalStorage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchLocalStorage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm: Added.
(testFetchLocalStorage):
(TEST(WebKit, FetchLocalStorageFromPersistentDataStore)):
(TEST(WebKit, FetchLocalStorageFromEphemeralDataStore)):
(-[RestoreLocalStorageMessageHandler userContentController:didReceiveScriptMessage:]):
(postMessage):
(setItem):
(FetchLocalStorageFromPersistentDataStoreThirdPartyIFrame)):
(FetchLocalStorageFromEphemeralDataStoreThirdPartyIFrame)):

Canonical link: https://commits.webkit.org/287349@main
